hazelcast-sql/src/main/java/com/hazelcast/sql/impl/expression/UntrustedExpressionEvalContext.java

```diff
@@ -0,0 +1,111 @@
+import javax.annotation.Nullable;
+import javax.security.auth.Subject;
+import java.security.Permission;
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Specifies an untrusted context for assessing expressions.
+ * Contains identical information as the trusted context {@code ExpressionEvalContextImpl}, except for the security context.
+ * If an attempt is made to acquire a security context, an exception will be raised,
+ * indicating that the context lacks trust.
+ * This class is utilized in scenarios where obtaining a trusted security context is unfeasible.
+ */
+public final class UntrustedExpressionEvalContext
+        implements ExpressionEvalContext {
+    private final List<Object> arguments;
+    private final transient InternalSerializationService serializationService;
+    private final transient NodeEngine nodeEngine;
+
+    public UntrustedExpressionEvalContext(
+            @Nonnull List<Object> arguments,
+            @Nonnull InternalSerializationService serializationService,
+            @Nonnull NodeEngine nodeEngine) {
+        this.arguments = requireNonNull(arguments);
+        this.serializationService = requireNonNull(serializationService);
+        this.nodeEngine = requireNonNull(nodeEngine);
+    }
+
+    public static UntrustedExpressionEvalContext from(ExpressionEvalContext context) {
+        return new UntrustedExpressionEvalContext(
+                context.getArguments(),
+                context.getSerializationService(),
+                context.getNodeEngine()
+        );
+    }
+
+    @Override
+    public Object getArgument(int index) {
+        return arguments.get(index);
+    }
+
+    @Override
+    public List<Object> getArguments() {
+        return arguments;
+    }
+
+    @Override
+    public InternalSerializationService getSerializationService() {
+        return serializationService;
+    }
+
+    @Override
+    public UntrustedExpressionEvalContext withSerializationService(@Nonnull InternalSerializationService newService) {
+        if (serializationService == newService) {
+            return this;
+        }
+        return new UntrustedExpressionEvalContext(
+                arguments,
+                newService,
+                nodeEngine
+        );
+    }
+
+    @Override
+    public NodeEngine getNodeEngine() {
+        return nodeEngine;
+    }
+
+    @Override
+    public void checkPermission(Permission permission) {
+        if (isSecurityEnabled()) {
+            throw new SecurityException("Unable to employ sensitive functions in untrusted invocations.");
+        }
+    }
+
+    @Override
+    @Nullable
+    public Subject subject() {
+        if (isSecurityEnabled()) {
+            throw new SecurityException("Unable to employ sensitive functions in untrusted invocations.");
+        }
+        return null;
+    }
+
+    private boolean isSecurityEnabled() {
+        return nodeEngine.getConfig().getSecurityConfig().isEnabled();
```

Comment:
It seems we can trust the node engine since it is configured after deserialization on the server side.

Reply:
yes. we cannot trust data that were sent over the wire. but data injected after deserialization is trusted (OFC if it does not depend on untrusted data).
Also note that EEC is not serialized directly. It is created during deserialzation of other objects from raw data.